File: JavaStudy/src/com/xzq/jvm/GcWatcher.java
package com.xzq.jvm;

import java.lang.ref.WeakReference;

public class GcWatcher {

    static WeakReference<GcOwner> weakReference = new WeakReference<>(new GcOwner());

    public static void main(String[] args) {

      /*
        Runtime runtime = Runtime.getRuntime();
        System.out.println("maxMemory=" + runtime.maxMemory() / (1024 * 1024));
        System.out.println("totalMemory=" + runtime.totalMemory() / (1024 * 1024));
        System.out.println("freeMemory=" + runtime.freeMemory() / (1024 * 1024));

        GcOwner gcOwner = new GcOwner();
        Runtime.getRuntime().gc();
        gcOwner = null;
        Runtime.getRuntime().gc();
        */

        System.out.println("------  GcOwner   finalize1 " + weakReference.get());
        Runtime.getRuntime().gc();
        System.out.println("------  GcOwner   finalize2 " + weakReference.get());
    }

}

class GcOwner {

    @Override
    protected void finalize() throws Throwable {
//        super.finalize();
        System.out.println("------  GcOwner   finalize");
    }
}

File: KotlinStudy/src/main/kotlin/com/xzq/jvm/GcWatcher.kt
package com.xzq.jvm

import java.lang.ref.WeakReference


object GcWatcher {

    @JvmStatic
    fun main(args: Array<String>) {
        println("-------")
    }


    private var sGcWatcher: WeakReference<GcWatcher>? = null
    private var sGcWatchers: ArrayList<Runnable?> = ArrayList()
    private var lock: Any = Any()
    private var sLastGcTime: Long = 0

    private class GcWatcher {
        @Throws(Throwable::class)
        protected fun finalize() {
            sLastGcTime = System.currentTimeMillis()
            var sTmpWatchers: ArrayList<Runnable?>
            synchronized(lock) {
                sTmpWatchers = sGcWatchers
                try {
                    sTmpWatchers.filterNotNull().forEach {
                        it.run()
                    }
//                    for (i in sTmpWatchers.indices) {
//                        if (sTmpWatchers[i] != null) {
//                            sTmpWatchers[i]!!.run()
//                        }
//                    }
                } catch (e: Throwable) {
                    e.printStackTrace()
                }
                sGcWatcher = WeakReference(GcWatcher())
            }
        }
    }

    fun addGcWatcher(watcher: Runnable?) {
        synchronized(lock) {
            sGcWatchers.add(watcher)
            if (sGcWatcher == null) {
                sGcWatcher = WeakReference(GcWatcher())
            }
        }
    }

    fun removeGcWatcher(watcher: Runnable?) {
        synchronized(lock) {
            sGcWatchers.remove(watcher)
            if (sGcWatchers.isEmpty()) {
                sGcWatcher = null
            }
        }
    }
}